### ISSUE TITLE
Find and replace QA testing

### ISSUE THREAD

**Issue by mlewand** [ckeditor/ckeditor5#9964]
Here's the ticket where we'll put the notes from feature testing.

**Comment by mlewand** [comment]
There are some known missing pieces on current `i/9802` branch:

*   Checkboxes do not work and have no effect (will be resolved in #9880 / #9958 ).
*   There are some minor styling issues listed in #9959.
*   There's an issue where undoing after replacing text will bring back some highlights.

**Comment by LukaszGudel** [comment]
### Steps:
1. Enter any text that ends with `\` character in `Find in text` field.
2. Click "Find" button or press Enter.

### Result:
Error in the console:
```
utils.js:114 Uncaught SyntaxError: Invalid regular expression: /foo\/: \ at end of pattern
    at new RegExp (<anonymous>)
    at findByTextCallback (utils.js:114)
    at FindCommand.execute (findcommand.js:47)
    at FindCommand.<anonymous> (observablemixin.js:258)
    at FindCommand.fire (emittermixin.js:199)
    at FindCommand.<computed> [as execute] (observablemixin.js:262)
    at CommandCollection.execute (commandcollection.js:69)
    at ClassicEditor.execute (editor.js:292)
    at FindAndReplaceUI.<anonymous> (findandreplace.js:51)
    at FindAndReplaceUI.fire (emittermixin.js:199)
```

**Comment by FilipTokarski** [comment]
Same with other characters like for example `(`, `)`, `{`, `}`, `[`, `]`, `?`

**Comment by FilipTokarski** [comment]
If you enter `.` and click `find` the editor freezes.

**Comment by mlewand** [comment]
:point\_up: Ok so it boils down to the fact that we're not escaping regexps

**Comment by FilipTokarski** [comment]
1. Find a word that is somewhere inside a formatted text/paragraph
2. Replace it

**Result:** newly inserted text loses formatting. I would intuitively assume the new word will have the same formatting as before the replacement.
  

https://user-images.githubusercontent.com/34380544/123620351-6aeb2500-d80a-11eb-91e3-a69824434ec1.mp4

**Comment by LukaszGudel** [comment]
### Steps
1. Add `foo()` to the editor content.
2. Try to find `foo()`.

### Result:
It only matches `foo`.
![image](https://user-images.githubusercontent.com/58658048/123621257-4f344e80-d80b-11eb-94b4-e578983abced.png)

### Notes:
- Same with `foo[]` or probably many other texts with special regex characters
- I'm not sure if this is related to the first issue with unescaped characters

It feels like it will be hard to support both plain text searches like `foo()` and regex searches. In VSCode there is a button for toggling between plain text matches and regex matches. Maybe it's something that could be implemented as well as 3rd checkbox?

**Comment by FilipTokarski** [comment]
1. Open FR balloon
2. Inside the input - press `space` and click `Find`

**Result:** it finds every space in the document. I suppose it should ignore such search.
  

https://user-images.githubusercontent.com/34380544/123622483-8bb47a00-d80c-11eb-83ab-dc45164d51cb.mp4

**Comment by LukaszGudel** [comment]
### Steps:
1. Look for `foo||bar`.

### Result:
Crash with error in the console:
```js
findandreplaceediting.js:275 Uncaught TypeError: editor.editing.mapper.markerNameToElements is not a function or its return value is not iterable
    at findandreplaceediting.js:275
    at Document._callPostFixers (document.js:186)
    at View.change (view.js:492)
    at View._disableRendering (view.js:685)
    at Model.EditingController.listenTo.priority (editingcontroller.js:86)
    at Model.fire (emittermixin.js:199)
    at Model._runPendingChanges (model.js:848)
    at Model.change (model.js:184)
    at utils.js:48
    at Array.forEach (<anonymous>)
```

Seems like another issue with unescaped characters, but the error is different so I'm posting it as well.

**Comment by FilipTokarski** [comment]
In RTL languages the dropdown styling is broken:
![FR_rtl](https://user-images.githubusercontent.com/34380544/123623550-daaedf00-d80d-11eb-89c4-de14ddc111f6.png)

**Comment by FilipTokarski** [comment]
1. Switch to source mode

**Expected:** find & replace toolbar dropdown is inactive.

**Actual:** find & replace toolbar dropdown is active, you can open it and type some text, but nothing happens anyway.

**Comment by LukaszGudel** [comment]
### Steps:
1. Have content with some `foo` words.
2. Find the word `foo`.
3. Replace all with `foo`.
4. Try to find `foo` again without closing dropdown.

### Result:
After replacing a word with the same word it's not possible to find it again without resetting find&replace dropdown.

**Comment by LukaszGudel** [comment]
### Steps: 
1. Find any word.
2. Replace it with some text.
3. Close F&R dropdown and undo.

### Result:
The highlight will stay on searched word, even when searching other words.

**Comment by FilipTokarski** [comment]
1. Switch to read-only mode

**Expected:** find & replace toolbar dropdown is inactive.

**Actual:** find & replace toolbar dropdown is active, you can open it and type some text, but nothing happens anyway.

**Comment by FilipTokarski** [comment]
How should the plugin behave in restricted editing mode? Currently, if you have some editable region and find a word that is inside it, you cannot change it with Find&Replace, even though you can edit it manually.

**Comment by LukaszGudel** [comment]
### Steps:
1. Search for some word.
2. Type some other word in search input.

### Result:
Previous search results (eg. `1 of 3`) is still visible and might be misleading, because it will not be correct for other text typed into the search input.

https://user-images.githubusercontent.com/58658048/123755891-ab0ae000-d8bc-11eb-86c8-aa17e562fa4a.mp4

**Comment by FilipTokarski** [comment]
**Steps:**
1. Find a word that occurs a few times in the editor
2. Click `REPLACE` a few times

**Expected:**
The word is replaced, focus stays in the same place. To go to the next found word you should click the arrow.

**Actual:**
When you replace the word, editor scrolls automatically to the next one, which creates this effect of sudden _jumping_.


https://user-images.githubusercontent.com/34380544/123756313-15238500-d8bd-11eb-986b-931e8b03c358.mp4

**Comment by LukaszGudel** [comment]
Find and replace dropdown styles are broken on mobile:

![image](https://user-images.githubusercontent.com/58658048/123782079-355f3e00-d8d5-11eb-8d5f-50b396a836d3.png)

- This screenshot is from chrome dev tools but it looks the same on my Xiaomi Mi6 with Android 9. 
- It's very similar on iPhones and Safari as well.
- dropdown on Safari on iPad looks all right (same as on desktop)
- screenshot is from docs but the same case is on manual tests so it's not some additional css rules from our docs

**Comment by LukaszGudel** [comment]
Seems like scrolling to the next result can also change the position of the editor on the page (page scroll). This might lead to the case where during spam clicking the next result button it can change the position of the dropdown and user will be clicking different button or close the dropdown by clicking inside the editor:


https://user-images.githubusercontent.com/58658048/123791937-59744c80-d8e0-11eb-93e0-cb7f02ad211c.mp4

**Comment by FilipTokarski** [comment]
The second input is not utilising its full width:
  

https://user-images.githubusercontent.com/34380544/123794877-a60d5700-d8e3-11eb-80c6-515889be6ef4.mp4

**Comment by LukaszGudel** [comment]
### [Safari only]

There is some kind of blue border/outline when focusing dropdown checkboxes on Safari:

https://user-images.githubusercontent.com/58658048/123795407-48c5d580-d8e4-11eb-8129-dc4d50a649c9.mp4

Recorded on `i/9959` but it's the same case on `i/9802`. Other browser do not have this outline when clicking on checkboxes.

**Comment by mlewand** [comment]
QA testing is complete. The most urging issues were either fixed or added to follow-ups in #10024.
